File: frontend/services/tools.ts
import { authenticatedGet } from "@/services/auth";
import type {
  ToolExecutionDetail,
  ToolExecutionListResponse,
  ToolListResponse,
} from "@/types/api";

type ToolsResult<T> =
  | { ok: true; data: T; status: number }
  | { ok: false; error: string; status: number | null };

export async function listTools(): Promise<ToolsResult<ToolListResponse>> {
  return authenticatedGet<ToolListResponse>("/api/v1/tools");
}

export async function listToolExecutions(params?: {
  limit?: number;
  offset?: number;
  conversation_id?: string;
}): Promise<ToolsResult<ToolExecutionListResponse>> {
  const search = new URLSearchParams();
  if (params?.limit != null) search.set("limit", String(params.limit));
  if (params?.offset != null) search.set("offset", String(params.offset));
  if (params?.conversation_id) search.set("conversation_id", params.conversation_id);
  const qs = search.toString();
  return authenticatedGet<ToolExecutionListResponse>(
    `/api/v1/tool-executions${qs ? `?${qs}` : ""}`,
  );
}

export async function getToolExecution(
  executionId: string,
): Promise<ToolsResult<ToolExecutionDetail>> {
  return authenticatedGet<ToolExecutionDetail>(`/api/v1/tool-executions/${executionId}`);
}

File: backend/app/api/routes/tools.py
"""Authenticated tool discovery and execution history APIs."""

from __future__ import annotations

import uuid
from typing import Annotated

from fastapi import APIRouter, Query

from app.api.deps import CurrentActiveUser, DbSessionDep, ToolServiceDep
from app.tools.schemas import (
    ToolExecutionDetail,
    ToolExecutionListResponse,
    ToolListResponse,
)

router = APIRouter(tags=["tools"])


@router.get(
    "/tools",
    response_model=ToolListResponse,
    summary="List tools available to the current user",
)
async def list_tools(
    user: CurrentActiveUser,
    tools: ToolServiceDep,
) -> ToolListResponse:
    return tools.list_tools_for_user(user)


@router.get(
    "/tool-executions",
    response_model=ToolExecutionListResponse,
    summary="List the current user's tool executions",
)
async def list_tool_executions(
    session: DbSessionDep,
    user: CurrentActiveUser,
    tools: ToolServiceDep,
    limit: Annotated[int, Query(ge=1, le=100)] = 20,
    offset: Annotated[int, Query(ge=0)] = 0,
    conversation_id: uuid.UUID | None = None,
) -> ToolExecutionListResponse:
    return await tools.list_executions(
        session,
        user,
        limit=limit,
        offset=offset,
        conversation_id=conversation_id,
    )


@router.get(
    "/tool-executions/{execution_id}",
    response_model=ToolExecutionDetail,
    summary="Get a tool execution owned by the current user",
)
async def get_tool_execution(
    execution_id: uuid.UUID,
    session: DbSessionDep,
    user: CurrentActiveUser,
    tools: ToolServiceDep,
) -> ToolExecutionDetail:
    return await tools.get_execution(session, user, execution_id)
